mlir/lib/Dialect/GPU/IR/GPUDialect.cpp

```diff
@@ -401,8 +401,29 @@ LogicalResult GPUDialect::verifyOperationAttribute(Operation *op,
              << expectedNumArguments;
 
     auto functionType = kernelGPUFunction.getFunctionType();
+    auto typesMatch = [&](Type launchOpArgType, Type gpuFuncArgType) {
+      auto launchOpMemref = dyn_cast<MemRefType>(launchOpArgType);
+      auto kernelMemref = dyn_cast<MemRefType>(gpuFuncArgType);
+      // Allow address space incompatibility for OpenCL kernels: `gpu.launch`'s
+      // argument memref without address space attribute will match a kernel
+      // function's memref argument with address space `Global`.
+      if (launchOpMemref && kernelMemref) {
+        auto launchAS = llvm::dyn_cast_or_null<gpu::AddressSpaceAttr>(
+            launchOpMemref.getMemorySpace());
+        auto kernelAS = llvm::dyn_cast_or_null<gpu::AddressSpaceAttr>(
+            kernelMemref.getMemorySpace());
+        if (!launchAS && kernelAS &&
+            kernelAS.getValue() == gpu::AddressSpace::Global)
```

Comment:
Can't you also pass a pointer to workgroup (aka local, shared) memory here from the host? Can we simply not check memory space at all between launchpoint and kernel?

Reply:
Yup, that's what I started with. I wasn't sure how specific the check should be, so just put this as an example. We can either go with skipping them altogether or describe the cases precisely.

Reply:
> Can't you also pass a pointer to workgroup (aka local, shared) memory here from the host?

Not sure, what would it mean? I.e. what's the expected behavior?

Reply:
According to the OpenCL spec  https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#setting-kernel-arguments:

```OpenCL
// Provided by CL_VERSION_1_0
cl_int clSetKernelArg(
    cl_kernel kernel,
    cl_uint arg_index,
    size_t arg_size,
    const void* arg_value);
```

> arg_size specifies the size of the argument value. If the argument is a memory object, the arg_size value must be equal to sizeof(cl_mem). *For arguments declared with the local qualifier, the size specified will be the size in bytes of the buffer that must be allocated for the local argument.* If the argument is of type sampler_t, the arg_size value must be equal to sizeof(cl_sampler). If the argument is of type queue_t, the arg_size value must be equal to sizeof(cl_command_queue). For all other arguments, the size will be the size of argument type.


Also, in the OpenCL-SPIR-V client specification https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Env.html#_kernel_arguments:

> For OpTypePointer parameters, supported Storage Classes are:
> - CrossWorkgroup
> - Workgroup
> - UniformConstant


I would also allow SLM pointer arguments here, as that's how, e.g., SYCL is doing it AFAIL

Reply:
Thanks for the links!

> According to the OpenCL spec https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#setting-kernel-arguments

I mean the SLM pointer passed from host would be `null` (and having just one pointer is meaningless anyway since each WG would have its own); this is also in the spec, same section:
> If the argument is declared with the local qualifier, the arg_value entry must be NULL.

We can allow it, but I don't see any practical sense. If a kernel has a parameter marked as local, then the caller should already know about it. In other words, the memory pointer (let's forget that it is NULL for a moment) that is passed to the kernel does not come from a runtime-managed buffer that can migrate between devices. Hence, no such problem should exist for those.

Current definition of the address space enum:
```
def GPU_AddressSpaceEnum : GPU_I32Enum<
  "AddressSpace", "GPU address space", [
    GPU_AddressSpaceGlobal,
    GPU_AddressSpaceWorkgroup,
    GPU_AddressSpacePrivate
  ]>;
```
I suppose for constants it should still have address space global and retrieve constness in the lowering?

P.s. the only reason I'm suggesting to make the check specific at all is that the check can actually be useful I assume, so it seems natural to try and preserve it for all the other cases.

Reply:
I see. In fact, I was thinking on a nonexistent issue. I think we're fine, as, when the kernel is still a `gpu.func`, local memory should be represented as memory attributions in the function, so, instead of:
```
gpu.func @kernel(%slm: memref<f32, #gpu.address_space<workgroup>>, ...)
```
we'll have:
```
gpu.func @kernel(...)
    workgroup(%slm: memref<f32, #gpu.address_space<workgroup>>)
```
And only when converting to LLMV dialect (no signature check anymore), we'll have:
```
llvm.func @kernel(..., %slm: llvm.ptr<3>)
```

TL;DR: We are fine. Thanks!